File: sdk/php/tests/SigningTest.php
<?php

declare(strict_types=1);

namespace Cordum\Cap\Tests;

use Cordum\Agent\V1\BusPacket;
use Cordum\Agent\V1\Heartbeat;
use Cordum\Agent\V1\JobRequest;
use Cordum\Cap\Codec;
use Cordum\Cap\Signing;
use Google\Protobuf\Timestamp;
use OpenSSLAsymmetricKey;
use PHPUnit\Framework\TestCase;

class SigningTest extends TestCase
{
    private function makePacket(): BusPacket
    {
        $req = new JobRequest();
        $req->setJobId('job-sign-1');
        $req->setTopic('sign.test');

        $ts = new Timestamp();
        $ts->setSeconds(1704164645);

        $pkt = new BusPacket();
        $pkt->setTraceId('trace-sign');
        $pkt->setSenderId('sender-sign');
        $pkt->setProtocolVersion(1);
        $pkt->setCreatedAt($ts);
        $pkt->setJobRequest($req);

        return $pkt;
    }

    /** @return array{0: OpenSSLAsymmetricKey, 1: OpenSSLAsymmetricKey} */
    private function makeKeyPair(): array
    {
        $key = openssl_pkey_new([
            'curve_name' => 'prime256v1',
            'private_key_type' => OPENSSL_KEYTYPE_EC,
        ]);
        $details = openssl_pkey_get_details($key);
        openssl_pkey_export($key, $privatePem);
        $publicPem = $details['key'];

        $priv = openssl_pkey_get_private($privatePem);
        $pub = openssl_pkey_get_public($publicPem);

        return [$priv, $pub];
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function signAndVerifyRoundTrip(): void
    {
        [$priv, $pub] = $this->makeKeyPair();
        $packet = $this->makePacket();

        Signing::signPacket($packet, $priv);
        $this->assertNotEmpty($packet->getSignature());
        $this->assertTrue(Signing::verifyPacketSignature($packet, $pub));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function signAndVerifyHeartbeat(): void
    {
        [$priv, $pub] = $this->makeKeyPair();

        $hb = new Heartbeat();
        $hb->setWorkerId('worker-hb');
        $hb->setPool('gpu');
        $hb->setActiveJobs(3);

        $pkt = new BusPacket();
        $pkt->setTraceId('trace-hb');
        $pkt->setSenderId('worker-hb');
        $pkt->setProtocolVersion(1);
        $pkt->setHeartbeat($hb);

        Signing::signPacket($pkt, $priv);
        $this->assertTrue(Signing::verifyPacketSignature($pkt, $pub));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function tamperTraceIdDetected(): void
    {
        [$priv, $pub] = $this->makeKeyPair();
        $packet = $this->makePacket();
        Signing::signPacket($packet, $priv);

        $packet->setTraceId('tampered');
        $this->assertFalse(Signing::verifyPacketSignature($packet, $pub));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function tamperSenderIdDetected(): void
    {
        [$priv, $pub] = $this->makeKeyPair();
        $packet = $this->makePacket();
        Signing::signPacket($packet, $priv);

        $packet->setSenderId('tampered');
        $this->assertFalse(Signing::verifyPacketSignature($packet, $pub));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function tamperPayloadDetected(): void
    {
        [$priv, $pub] = $this->makeKeyPair();
        $packet = $this->makePacket();
        Signing::signPacket($packet, $priv);

        $tampered = new JobRequest();
        $tampered->setJobId('tampered');
        $tampered->setTopic('tampered');
        $packet->setJobRequest($tampered);
        $this->assertFalse(Signing::verifyPacketSignature($packet, $pub));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptySignatureRejected(): void
    {
        [, $pub] = $this->makeKeyPair();
        $packet = $this->makePacket();
        $this->assertFalse(Signing::verifyPacketSignature($packet, $pub));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function wrongKeyRejected(): void
    {
        [$priv1,] = $this->makeKeyPair();
        [, $pub2] = $this->makeKeyPair();

        $packet = $this->makePacket();
        Signing::signPacket($packet, $priv1);
        $this->assertFalse(Signing::verifyPacketSignature($packet, $pub2));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function signatureIsDerEncoded(): void
    {
        [$priv,] = $this->makeKeyPair();
        $packet = $this->makePacket();
        Signing::signPacket($packet, $priv);

        $sig = $packet->getSignature();
        // DER-encoded ECDSA signatures start with 0x30 (SEQUENCE tag)
        $this->assertSame(0x30, ord($sig[0]));
        // P-256 DER signatures are typically 68-73 bytes
        $this->assertGreaterThanOrEqual(68, strlen($sig));
        $this->assertLessThanOrEqual(73, strlen($sig));
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function signingDoesNotAlterContent(): void
    {
        [$priv,] = $this->makeKeyPair();
        $packet = $this->makePacket();
        $clean = clone $packet;
        $clean->setSignature('');
        $originalBytes = Codec::marshalDeterministic($clean);

        Signing::signPacket($packet, $priv);

        $after = clone $packet;
        $after->setSignature('');
        $afterBytes = Codec::marshalDeterministic($after);

        $this->assertSame($originalBytes, $afterBytes);
    }
}

File: sdk/rust/tests/signing_test.rs
use cap_sdk::codec::marshal_deterministic;
use cap_sdk::pb::{bus_packet::Payload, BusPacket, Heartbeat, JobRequest, JobResult};
use cap_sdk::signing::{sign_packet, verify_packet_signature};
use p256::ecdsa::{SigningKey, VerifyingKey};
use rand_core::OsRng;

fn make_packet() -> BusPacket {
    BusPacket {
        trace_id: "trace-sign".into(),
        sender_id: "sender-sign".into(),
        protocol_version: 1,
        created_at: Some(prost_types::Timestamp {
            seconds: 1704164645,
            nanos: 0,
        }),
        payload: Some(Payload::JobRequest(JobRequest {
            job_id: "job-sign-1".into(),
            topic: "sign.test".into(),
            ..Default::default()
        })),
        ..Default::default()
    }
}

fn make_key_pair() -> (SigningKey, VerifyingKey) {
    let sk = SigningKey::random(&mut OsRng);
    let vk = VerifyingKey::from(&sk);
    (sk, vk)
}

#[test]
fn sign_and_verify_job_request() {
    let (sk, vk) = make_key_pair();
    let mut packet = make_packet();

    sign_packet(&mut packet, &sk).unwrap();
    assert!(!packet.signature.is_empty());

    verify_packet_signature(&packet, &vk).unwrap();
}

#[test]
fn sign_and_verify_heartbeat() {
    let (sk, vk) = make_key_pair();
    let mut packet = BusPacket {
        trace_id: "trace-hb".into(),
        sender_id: "worker-hb".into(),
        protocol_version: 1,
        payload: Some(Payload::Heartbeat(Heartbeat {
            worker_id: "worker-hb".into(),
            pool: "gpu".into(),
            active_jobs: 3,
            max_parallel_jobs: 10,
            cpu_load: 75.0,
            ..Default::default()
        })),
        ..Default::default()
    };

    sign_packet(&mut packet, &sk).unwrap();
    verify_packet_signature(&packet, &vk).unwrap();
}

#[test]
fn sign_and_verify_job_result() {
    let (sk, vk) = make_key_pair();
    let mut packet = BusPacket {
        trace_id: "trace-res".into(),
        sender_id: "worker-1".into(),
        protocol_version: 1,
        payload: Some(Payload::JobResult(JobResult {
            job_id: "job-res-1".into(),
            worker_id: "worker-1".into(),
            status: 5,
            ..Default::default()
        })),
        ..Default::default()
    };

    sign_packet(&mut packet, &sk).unwrap();
    verify_packet_signature(&packet, &vk).unwrap();
}

#[test]
fn tamper_trace_id_detected() {
    let (sk, vk) = make_key_pair();
    let mut packet = make_packet();
    sign_packet(&mut packet, &sk).unwrap();

    packet.trace_id = "tampered-trace".into();
    assert!(verify_packet_signature(&packet, &vk).is_err());
}

#[test]
fn tamper_sender_id_detected() {
    let (sk, vk) = make_key_pair();
    let mut packet = make_packet();
    sign_packet(&mut packet, &sk).unwrap();

    packet.sender_id = "tampered-sender".into();
    assert!(verify_packet_signature(&packet, &vk).is_err());
}

#[test]
fn tamper_payload_detected() {
    let (sk, vk) = make_key_pair();
    let mut packet = make_packet();
    sign_packet(&mut packet, &sk).unwrap();

    packet.payload = Some(Payload::JobRequest(JobRequest {
        job_id: "tampered-job".into(),
        topic: "tampered.topic".into(),
        ..Default::default()
    }));
    assert!(verify_packet_signature(&packet, &vk).is_err());
}

#[test]
fn tamper_protocol_version_detected() {
    let (sk, vk) = make_key_pair();
    let mut packet = make_packet();
    sign_packet(&mut packet, &sk).unwrap();

    packet.protocol_version = 99;
    assert!(verify_packet_signature(&packet, &vk).is_err());
}

#[test]
fn empty_signature_rejected() {
    let (_, vk) = make_key_pair();
    let packet = make_packet();
    assert!(verify_packet_signature(&packet, &vk).is_err());
}

#[test]
fn garbage_signature_rejected() {
    let (_, vk) = make_key_pair();
    let mut packet = make_packet();
    packet.signature = vec![0xFF, 0xFE, 0xFD, 0xFC];
    assert!(verify_packet_signature(&packet, &vk).is_err());
}

#[test]
fn wrong_key_rejected() {
    let (sk1, _) = make_key_pair();
    let (_, vk2) = make_key_pair();

    let mut packet = make_packet();
    sign_packet(&mut packet, &sk1).unwrap();

    assert!(verify_packet_signature(&packet, &vk2).is_err());
}

#[test]
fn signature_is_der_encoded() {
    let (sk, _) = make_key_pair();
    let mut packet = make_packet();
    sign_packet(&mut packet, &sk).unwrap();

    // DER-encoded ECDSA signatures start with 0x30 (SEQUENCE tag)
    assert_eq!(packet.signature[0], 0x30);
    // P-256 DER signatures are typically 70-72 bytes
    assert!(
        packet.signature.len() >= 68 && packet.signature.len() <= 73,
        "DER signature length {} outside expected range",
        packet.signature.len()
    );
}

#[test]
fn signing_does_not_alter_packet_content() {
    let (sk, _) = make_key_pair();
    let mut packet = make_packet();
    let original_bytes = marshal_deterministic(&{
        let mut p = packet.clone();
        p.signature = Vec::new();
        p
    });

    sign_packet(&mut packet, &sk).unwrap();

    // Strip signature and re-serialize — should match original
    let mut after = packet.clone();
    after.signature = Vec::new();
    let after_bytes = marshal_deterministic(&after);

    assert_eq!(original_bytes, after_bytes);
}

#[test]
fn double_sign_overwrites() {
    let (sk, vk) = make_key_pair();
    let mut packet = make_packet();

    sign_packet(&mut packet, &sk).unwrap();
    let first_sig = packet.signature.clone();

    sign_packet(&mut packet, &sk).unwrap();
    let second_sig = packet.signature.clone();

    // Both signatures should verify
    verify_packet_signature(&packet, &vk).unwrap();

    // Signatures may differ (ECDSA uses a nonce) but both are valid
    // Just ensure the second sign didn't break anything
    assert!(!first_sig.is_empty());
    assert!(!second_sig.is_empty());
}
